### PR TITLE
Db misc

### PR DESCRIPTION
**Pull request by AndreaLanfranchi**
* Remove ambiguity on datapath passed by command line (whether it is a file or a directory)
* Introduce `DataDirectory` class which holds and manage correct paths (actually calling method parent() on datadir returns different result whether original input has a trailing slash or not - eg. parent of c:\test\data\ is still c:\test\data while parent of c:\test\data is c:\test)
* Introduce db::for_each
* Introduce db::for_count

### CODE REVIEW

File: core/silkworm/common/util.cpp [hunk 1]
```diff
@@ -231,4 +226,18 @@ size_t prefix_length(ByteView a, ByteView b) {
     }
     return len;
 }
+
+std::vector<std::string> split(std::string source, std::string delimiter) {
```
Comment (yperbasis) [inline core/silkworm/common/util.cpp:230]: I know it's trivial, but could you please add a test case?

File: db/silkworm/db/mdbx.cpp [hunk 2]
```diff
@@ -28,6 +27,9 @@ ::mdbx::env_managed open_env(const EnvConfig& config) {
 
     // Check datafile exists if create is not set
     fs::path db_path{config.path};
+    if (db_path.has_filename()) {
+        db_path += std::filesystem::path::preferred_separator;  // Remove amibiguity. It hash to be a directory
```
Comment (yperbasis) [inline db/silkworm/db/mdbx.cpp:31]: "hash" typo

**Review by yperbasis — APPROVED** [review]

File: core/silkworm/common/util.cpp (hunk 1 above)
Comment (AndreaLanfranchi) [reply core/silkworm/common/util.cpp:230]: Yeh right. Which also made emerge a couple of cases I did not think of.
Thank you

File: db/silkworm/db/mdbx.cpp (hunk 2 above)
Comment (AndreaLanfranchi) [reply db/silkworm/db/mdbx.cpp:31]: And "amibiguity" too

**Pull request merged**